packages/gatsby/src/recipes/graphql.js

```diff
@@ -0,0 +1,284 @@
+  fields: {
+    name: { type: GraphQLString },
+    version: { type: GraphQLString },
+  },
+})
+
+const rootQueryType = new GraphQLObjectType({
+  name: `Root`,
+  fields: () => {
+    return {
+      npmPackage: {
+        type: NPMPackageType,
+        args: {
+          name: {
+            type: GraphQLString,
+          },
+        },
+        resolve: async (_, args) => {
+          // TODO: peer/dev
+          const { dependencies } = await readPackage()
+          const version = dependencies[args.name]
+          return { name: args.name, version }
+        },
+      },
+      allNpmPackage: {
+        type: new GraphQLList(NPMPackageType),
+        resolve: async () => {
+          const { dependencies } = await readPackage()
+
+          return Object.entries(dependencies).map(([name, version]) => {
+            return {
+              name,
+              version,
+            }
+          })
+        },
+      },
+      allNpmPackageScripts: {
+        type: new GraphQLList(NPMPackageScriptType),
+        resolve: async () => {
+          const { scripts } = await readPackage()
+          return Object.entries(scripts).map(([name, script]) => {
+            return {
+              name,
+              script,
+            }
+          })
+        },
+      },
+      allGatsbyConfigPlugin: {
+        type: new GraphQLList(GatsbyConfigPluginType),
+        resolve: async () => {
+          const plugins = gatsbyPluginResource.read({ root })
+          return plugins.map(async name => {
+            return { name }
+          })
+        },
+      },
+    }
+  },
+})
+
+const rootMutationType = new GraphQLObjectType({
+  name: `Mutation`,
+  fields: () => {
+    return {
+      createOperation: {
+        type: GraphQLString,
+        args: {
+          commands: { type: GraphQLString },
+        },
+        resolve: (_data, args) => {
+          applyPlan(JSON.parse(args.commands))
+        },
+      },
+    }
+  },
+})
+
+const rootSubscriptionType = new GraphQLObjectType({
+  name: `Subscription`,
+  fields: () => {
+    return {
+      operation: {
+        type: OperationType,
+        subscribe: () => pubsub.asyncIterator(`operation`),
+        resolve: payload => payload,
+      },
+      npmPackageUpdated: {
+        type: NPMPackageType,
+        args: {
+          name: {
+            type: GraphQLString,
+          },
+        },
+        subscribe: withFilter(
+          () => pubsub.asyncIterator(`npmPackageUpdated`),
+          (payload, variables) =>
+            true || payload.npmPackageUpdated.dependencies[variables.name]
+        ),
+        resolve: () => {
+          return {
+            version: `${Math.random()}`,
+          }
+        },
+      },
+    }
+  },
+})
+
+const schema = new GraphQLSchema({
+  query: rootQueryType,
+  mutation: rootMutationType,
+  subscription: rootSubscriptionType,
+})
+
+const app = express()
+const server = createServer(app)
+
+console.log(`listening on localhost:4000`)
```

Comment:
Port should be customizable and probably should check if that port is free.

(This all kinda makes me question why we need a separate server to run all this)

Reply:
yeah we're doing this — btw, check out the TODOs https://github.com/gatsbyjs/gatsby/pull/22709/files#diff-107dedc914ad12f4801c053d2b306e40